activerecord/lib/active_record/associations/preloader/association.rb

```diff
@@ -4,16 +4,52 @@ module ActiveRecord
   module Associations
     class Preloader
       class Association #:nodoc:
-        def self.load_records_in_batch(scope, association_key_name, loaders)
-          ids = loaders.flat_map(&:owner_keys).uniq
+        class LoaderQuery
+          attr_reader :scope, :association_key_name
 
-          raw_records = scope.where(association_key_name => ids).load do |record|
-            loaders.each { |l| l.set_inverse(record) }
+          def initialize(scope, association_key_name)
+            @scope = scope
+            @association_key_name = association_key_name
           end
 
-          loaders.each do |loader|
-            loader.load_records(raw_records)
-            loader.run
+          def eql?(other)
+            return false unless association_key_name == other.association_key_name
+
+            # Attempt to inexpensively rule out or confirm equality
+            return false unless scope.table_name == other.scope.table_name
+            return false unless scope.preload_values == other.scope.preload_values
+            return false unless scope.includes_values == other.scope.includes_values
+            return true if scope.empty_scope? && other.scope.empty_scope?
+            return false unless scope.where_values_hash == other.scope.where_values_hash
+            return false unless scope.values.empty? == other.scope.values.empty?
+
+            # Pretty good chance of equality, let's fully compare SQL
+            scope.to_sql == other.scope.to_sql
+          end
+
+          # hash intentionally avoids scope.to_sql, so that it is only
+          # calculated when needed.
+          # This may cause extra collisions and thus `eql?` comparisons, but
+          # shouldn't be hugely significant.
+          def hash
+            [scope.table_name, scope.where_values_hash, association_key_name].hash
```

Comment:
`scope.where_values_hash` is filtered to include only Equality and In conditions, not Between or NotEqual conditions etc, recommends `scope.values` or `scope.where_clause` as an alternative.

Reply:
Unfortunately `WhereClause` and similar don't implement `hash` (they'll use `Object#hash` and basically hash by identity). This is why we use `scope.to_sql` above to compare, but I want to avoid needing to make that calculation in `hash`. This doesn't need to be exact (`eql?` does), just provide a good enough key that we don't have _too_ many collisions.

Reply:
How about adding `WhereClause#hash`?

```diff
diff --git a/activerecord/lib/active_record/relation/where_clause.rb b/activerecord/lib/active_record/relation/where_clause.rb
index 37fbb12575..f75710adcc 100644
--- a/activerecord/lib/active_record/relation/where_clause.rb
+++ b/activerecord/lib/active_record/relation/where_clause.rb
@@ -72,6 +72,10 @@ def ast
         predicates.one? ? predicates.first : Arel::Nodes::And.new(predicates)
       end
 
+      def hash
+        [self.class, predicates].hash
+      end
+
       def ==(other)
         other.is_a?(WhereClause) &&
           predicates == other.predicates
```

Reply:
Awesome! Didn't expect it to be that simple. I might be able to replace the rest of the `eql?` as well.